File: app/src/main/java/com/robifr/ledger/util/CurrencyFormat.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.util;

import androidx.annotation.NonNull;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.text.NumberFormat;
import java.text.ParseException;
import java.util.Locale;
import java.util.Objects;

public class CurrencyFormat {
  public static final int MINIMUM_FRACTION_DIGITS = 0;
  public static final int MAXIMUM_FRACTION_DIGITS = 5;

  private CurrencyFormat() {}

  public static String format(
      @NonNull BigDecimal amount, @NonNull String language, @NonNull String country) {
    return CurrencyFormat.format(
        amount, language, country, CurrencyFormat.symbol(language, country));
  }

  /**
   * @return Formatted amount into local specific currency.
   */
  @NonNull
  public static String format(
      @NonNull BigDecimal amount,
      @NonNull String language,
      @NonNull String country,
      @NonNull String symbol) {
    Objects.requireNonNull(amount);
    Objects.requireNonNull(language);
    Objects.requireNonNull(country);
    Objects.requireNonNull(symbol);

    final DecimalFormat format =
        (DecimalFormat) NumberFormat.getCurrencyInstance(new Locale(language, country));
    final DecimalFormatSymbols symbols = format.getDecimalFormatSymbols();

    symbols.setCurrencySymbol(symbol);
    format.setRoundingMode(RoundingMode.DOWN);
    format.setMinimumFractionDigits(CurrencyFormat.MINIMUM_FRACTION_DIGITS);
    format.setMaximumFractionDigits(CurrencyFormat.MAXIMUM_FRACTION_DIGITS);
    format.setDecimalFormatSymbols(symbols);
    return format.format(amount);
  }

  /** {@link CurrencyFormat#formatWithUnit(BigDecimal, String, String, String)} */
  @NonNull
  public static String formatWithUnit(
      @NonNull BigDecimal amount, @NonNull String language, @NonNull String country) {
    return CurrencyFormat.formatWithUnit(
        amount, language, country, CurrencyFormat.symbol(language, country));
  }

  /**
   * @return Formatted amount into local specific currency with an appropriate suffix (such as K for
   *     thousands or M for millions) appended at the end of the string.
   */
  @NonNull
  public static String formatWithUnit(
      @NonNull BigDecimal amount,
      @NonNull String language,
      @NonNull String country,
      @NonNull String symbol) {
    Objects.requireNonNull(amount);
    Objects.requireNonNull(language);
    Objects.requireNonNull(country);
    Objects.requireNonNull(symbol);

    final BigDecimal thousand = BigDecimal.valueOf(1000);
    final BigDecimal million = BigDecimal.valueOf(1_000_000);
    final BigDecimal billion = BigDecimal.valueOf(1_000_000_000);
    final BigDecimal trillion = BigDecimal.valueOf(1_000_000_000_000L);

    // Convert negative amount to positive to handle where negative amounts
    // can't be formatted due to division.
    final String negativePrefix = amount.compareTo(BigDecimal.ZERO) < 0 ? "-" : "";
    final BigDecimal positiveAmount = amount.abs();

    if (positiveAmount.compareTo(thousand) < 0) {
      return negativePrefix + CurrencyFormat.format(positiveAmount, language, country, symbol);

    } else if (positiveAmount.compareTo(million) < 0) {
      return negativePrefix
          + CurrencyFormat.format(
              positiveAmount.divide(thousand, 1, RoundingMode.DOWN), language, country, symbol)
          + "K";

    } else if (positiveAmount.compareTo(billion) < 0) {
      return negativePrefix
          + CurrencyFormat.format(
              positiveAmount.divide(million, 1, RoundingMode.DOWN), language, country, symbol)
          + "M";

    } else if (positiveAmount.compareTo(trillion) < 0) {
      return negativePrefix
          + CurrencyFormat.format(
              positiveAmount.divide(billion, 1, RoundingMode.DOWN), language, country, symbol)
          + "B";
    }

    return negativePrefix
        + CurrencyFormat.format(
            positiveAmount.divide(trillion, 1, RoundingMode.DOWN), language, country, symbol)
        + "T";
  }

  /**
   * @return Parsed amount from local specific currency.
   */
  @NonNull
  public static BigDecimal parse(
      @NonNull String amount, @NonNull String language, @NonNull String country)
      throws ParseException {
    Objects.requireNonNull(amount);
    Objects.requireNonNull(language);
    Objects.requireNonNull(country);

    final DecimalFormat format =
        (DecimalFormat) NumberFormat.getNumberInstance(new Locale(language, country));
    format.setParseBigDecimal(true);

    final String decimalSeparator = CurrencyFormat.decimalSeparator(language, country);
    String amountToParse = amount.replaceAll("[^\\d\\-\\" + decimalSeparator + "]", "");

    // Edge case.
    if (amountToParse.isBlank()
        || amountToParse.equals(decimalSeparator)
        || amountToParse.equals("-")
        || amountToParse.equals("-" + decimalSeparator)
        || Strings.countOccurrence(amountToParse, "-") > 1) {
      amountToParse = "0";
    }

    return ((BigDecimal) format.parse(amountToParse)).stripTrailingZeros();
  }

  @NonNull
  public static String symbol(@NonNull String language, @NonNull String country) {
    Objects.requireNonNull(language);
    Objects.requireNonNull(country);

    return new DecimalFormatSymbols(new Locale(language, country)).getCurrencySymbol();
  }

  @NonNull
  public static String groupingSeparator(@NonNull String language, @NonNull String country) {
    Objects.requireNonNull(language);
    Objects.requireNonNull(country);

    final char separator =
        new DecimalFormatSymbols(new Locale(language, country)).getGroupingSeparator();
    return Character.toString(separator);
  }

  @NonNull
  public static String decimalSeparator(@NonNull String language, @NonNull String country) {
    Objects.requireNonNull(language);
    Objects.requireNonNull(country);

    final char separator =
        new DecimalFormatSymbols(new Locale(language, country)).getDecimalSeparator();
    return Character.toString(separator);
  }

  public static int countDecimalPlace(@NonNull BigDecimal amount) {
    Objects.requireNonNull(amount);

    return Math.max(0, amount.stripTrailingZeros().scale());
  }

  public static boolean isSymbolAtStart(@NonNull String language, @NonNull String country) {
    Objects.requireNonNull(language);
    Objects.requireNonNull(country);

    final DecimalFormat format =
        (DecimalFormat) NumberFormat.getCurrencyInstance(new Locale(language, country));
    return format.toLocalizedPattern().indexOf('\u00A4') == 0;
  }
}

File: app/src/main/java/com/robifr/ledger/util/CurrencyFormat.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.util

import android.content.Context
import com.robifr.ledger.R
import java.math.BigDecimal
import java.math.RoundingMode
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.text.NumberFormat
import java.text.ParseException
import java.util.Locale
import kotlin.math.max

object CurrencyFormat {
  /**
   * Formats the given amount into a locale-specific currency string. Use [formatCents] to format
   * cents amount to its biggest currency unit.
   *
   * @param amount The amount to format.
   * @param languageTag The IETF BCP 47 language tag for locale formatting.
   * @param symbol The currency symbol to use.
   * @param fractionDigits The number of fraction digits to display.
   * @return A formatted currency string.
   */
  fun format(
      amount: BigDecimal,
      languageTag: String,
      symbol: String = symbol(languageTag),
      fractionDigits: Int = decimalFractionDigits(languageTag)
  ): String =
      (NumberFormat.getCurrencyInstance(Locale.forLanguageTag(languageTag)) as DecimalFormat)
          .apply {
            roundingMode = RoundingMode.DOWN
            minimumFractionDigits = 0
            maximumFractionDigits = fractionDigits
            // Directly modifying `currencySymbol` may not work due to immutability.
            decimalFormatSymbols = decimalFormatSymbols.apply { currencySymbol = symbol }
          }
          .format(amount)

  /**
   * Formats the given cents amount into a biggest locale-specific currency string. e.g. `1234`
   * becomes `"$12.34"`.
   *
   * @param amount The amount to format.
   * @param languageTag The IETF BCP 47 language tag for locale formatting.
   * @param symbol The currency symbol to use.
   * @return A formatted currency string.
   */
  fun formatCents(
      amount: BigDecimal,
      languageTag: String,
      symbol: String = symbol(languageTag)
  ): String = format(fromCents(amount, languageTag), languageTag, symbol)

  /**
   * Formats the given amount into a locale-specific currency string with an appropriate suffix
   * appended at the end of the string. e.g. "K" for thousands or "M" for millions).
   *
   * @param context The context for accessing resources or configuration.
   * @param amount The amount to format.
   * @param languageTag The IETF BCP 47 language tag for locale formatting.
   * @param symbol The currency symbol to use.
   * @param fractionDigits The number of fraction digits to display.
   * @return A formatted currency string.
   */
  fun formatWithUnit(
      context: Context,
      amount: BigDecimal,
      languageTag: String,
      symbol: String = symbol(languageTag),
      fractionDigits: Int = decimalFractionDigits(languageTag)
  ): String {
    val thousand: BigDecimal = 1000.toBigDecimal()
    val million: BigDecimal = 1_000_000.toBigDecimal()
    val billion: BigDecimal = 1_000_000_000.toBigDecimal()
    val trillion: BigDecimal = 1_000_000_000_000L.toBigDecimal()
    // Convert negative amount to positive to handle where negative amounts
    // can't be formatted due to division.
    val negativePrefix: String = if (amount.compareTo(0.toBigDecimal()) < 0) "-" else ""
    val positiveAmount: BigDecimal = amount.abs()
    return if (positiveAmount.compareTo(thousand) < 0) {
      negativePrefix + format(positiveAmount, languageTag, symbol, fractionDigits)
    } else if (positiveAmount.compareTo(million) < 0) {
      negativePrefix +
          format(
              positiveAmount.divide(thousand, 1, RoundingMode.DOWN),
              languageTag,
              symbol,
              fractionDigits) +
          context.getString(R.string.symbol_thousand)
    } else if (positiveAmount.compareTo(billion) < 0) {
      negativePrefix +
          format(
              positiveAmount.divide(million, 1, RoundingMode.DOWN),
              languageTag,
              symbol,
              fractionDigits) +
          context.getString(R.string.symbol_million)
    } else if (positiveAmount.compareTo(trillion) < 0) {
      negativePrefix +
          format(
              positiveAmount.divide(billion, 1, RoundingMode.DOWN),
              languageTag,
              symbol,
              fractionDigits) +
          context.getString(R.string.symbol_billion)
    } else {
      negativePrefix +
          format(
              positiveAmount.divide(trillion, 1, RoundingMode.DOWN),
              languageTag,
              symbol,
              fractionDigits) +
          context.getString(R.string.symbol_trillion)
    }
  }

  /**
   * Parses a locale-specific currency string into a numeric amount. Use [parseToCents] to parse
   * amount as cents.
   *
   * @param amount The currency string to parse.
   * @param languageTag The IETF BCP 47 language tag for locale parsing.
   * @param fractionDigits The number of fraction digits to display.
   * @return The parsed amount as a [BigDecimal].
   * @throws ParseException If the input string can't be parsed.
   */
  @Throws(ParseException::class)
  fun parse(
      amount: String,
      languageTag: String,
      fractionDigits: Int = decimalFractionDigits(languageTag)
  ): BigDecimal {
    val format: DecimalFormat =
        (NumberFormat.getNumberInstance(Locale.forLanguageTag(languageTag)) as DecimalFormat)
            .apply { isParseBigDecimal = true }
    // Replace every character except those that can be edited by the user
    // (digits, negative sign, and decimal separator).
    val amountToParse: String =
        amount
            .replace("""[^\d\-${Regex.escape(decimalSeparator(languageTag))}]""".toRegex(), "")
            .let { if (!isValidToParseAndFormat(amount, languageTag, fractionDigits)) "0" else it }
    return (format.parse(amountToParse) as BigDecimal).stripTrailingZeros()
  }

  /**
   * Parses a locale-specific currency string into a numeric cents amount. e.g. `"$123"` becomes
   * `12300`.
   *
   * @param amount The currency string to parse.
   * @param languageTag The IETF BCP 47 language tag for locale parsing.
   * @return The parsed amount as a [BigDecimal].
   * @throws ParseException If the input string can't be parsed.
   */
  @Throws(ParseException::class)
  fun parseToCents(amount: String, languageTag: String): BigDecimal =
      toCents(parse(amount, languageTag, decimalFractionDigits(languageTag)), languageTag)
          .stripTrailingZeros()

  /** Scales the given amount from cents to base units based on locale-specific fraction digits. */
  fun fromCents(
      amount: BigDecimal,
      languageTag: String,
      fractionDigits: Int = decimalFractionDigits(languageTag)
  ): BigDecimal =
      if (fractionDigits == -1) 0.toBigDecimal()
      else amount.divide(10.toBigDecimal().pow(fractionDigits), fractionDigits, RoundingMode.DOWN)

  /** Scales the given amount from base units to cents based on locale-specific fraction digits. */
  fun toCents(amount: BigDecimal, languageTag: String): BigDecimal {
    val fractionDigits: Int = decimalFractionDigits(languageTag)
    if (fractionDigits == -1) return 0.toBigDecimal()
    return 10.toBigDecimal().pow(fractionDigits) * amount
  }

  fun isValidToParseAndFormat(
      amount: String,
      languageTag: String,
      fractionDigits: Int = decimalFractionDigits(languageTag)
  ): Boolean {
    val decimalSeparator: String = decimalSeparator(languageTag)
    // Replace every character except those that can be edited by the user
    // (digits, negative sign, and decimal separator).
    val amount: String =
        amount.replace("""[^\d\-${Regex.escape(decimalSeparator)}]""".toRegex(), "")
    // Can't find any digit.
    return !(amount.count { it.isDigit() } == 0 ||
        // Found multiple decimal separator.
        amount.count { it.toString() == decimalSeparator } > 1 ||
        // Found multiple negative sign.
        amount.count { it.toString() == "-" } > 1 ||
        // Found any character before negative sign.
        """([^\-]+)(?=-)""".toRegex().findAll(amount).count() > 0 ||
        // Found the digit occurrence in decimal place is more than max allowed.
        ("""(?<=${Regex.escape(decimalSeparator)})\d+""".toRegex().find(amount)?.value?.length
            ?: 0) > fractionDigits)
  }

  fun symbol(languageTag: String): String =
      DecimalFormatSymbols(Locale.forLanguageTag(languageTag)).currencySymbol

  fun isSymbolAtStart(languageTag: String): Boolean =
      (NumberFormat.getCurrencyInstance(Locale.forLanguageTag(languageTag)) as DecimalFormat)
          .toLocalizedPattern()
          .indexOf('\u00A4') == 0

  fun groupingSeparator(languageTag: String): String =
      DecimalFormatSymbols(Locale.forLanguageTag(languageTag)).groupingSeparator.toString()

  fun decimalSeparator(languageTag: String): String =
      DecimalFormatSymbols(Locale.forLanguageTag(languageTag)).decimalSeparator.toString()

  fun decimalFractionDigits(languageTag: String): Int =
      NumberFormat.getCurrencyInstance(Locale.forLanguageTag(languageTag))
          .currency
          ?.defaultFractionDigits ?: -1

  fun countDecimalPlace(amount: BigDecimal): Int =
      max(0.0, amount.stripTrailingZeros().scale().toDouble()).toInt()
}
